pkg/proxy/backend/backend_conn_mgr.go

```diff
@@ -178,15 +179,31 @@ func (mgr *BackendConnManager) getBackendIO(ctx ConnContext, auth *Authenticator
 		return nil, err
 	}
 
-	mgr.logger.Info("found", zap.String("addr", addr))
-	mgr.backendConn = NewBackendConnection(addr)
-	if err := mgr.backendConn.Connect(); err != nil {
-		mgr.handshakeHandler.OnHandshake(auth, addr, err)
-		return nil, err
+	// Try to connect to all backup backends one by one until success.
+	maxRetries := 3
+	for retries := 0; retries < maxRetries; retries++ {
+		for len(addr) > 0 {
+			backendConn := NewBackendConnection(addr)
+			err = backendConn.Connect()
+			mgr.handshakeHandler.OnHandshake(auth, addr, err)
+			if err == nil {
+				if err := selector.Succeed(mgr); err == nil {
+					mgr.logger.Info("connected to backend", zap.String("addr", addr))
+					mgr.backendConn = backendConn
+					auth.serverAddr = addr
+					return mgr.backendConn.PacketIO(), nil
+				}
+				// Bad luck: the backend has been recycled or shut down just after the selector returns it.
+				if ignoredErr := backendConn.Close(); ignoredErr != nil {
+					mgr.logger.Error("close backend connection failed", zap.String("addr", addr), zap.Error(ignoredErr))
+				}
+			}
+			addr = selector.Next()
+		}
```

Comment:
I think we can integrate this loop with `backoff.RetryNotifyWithData` loop above. It is basically same thing with and without `Connect()`.

I think the following will be good:

```
for retries := 0; retries < maxRetries; retries++ {
   selector.Reset()
   backoff.backoff.RetryNotifyWithData(
		func() (*pnet.PacketIO, error) {
   ...
}
```

Or maybe the following, then backoff is more controlled by time instead of retry times:

```
selector := r.Route()
backoff.backoff.RetryNotifyWithData(
		func() (*pnet.PacketIO, error) {
                    addr := selector.Next()
                    if len(addr) == 0 { selector.Reset(); addr =  selector.Next() }
```